# Gowalla - Modelling
After looking at the data in the EDA, it's time to build our prediction model. The main goal is to predict if two users will be friends based on their behavior and their place in the network.

## Initialization

In [1]:
#Imports
import pandas as pd
import networkx as nx
import math
import random
import numpy as np

from haversine import haversine_vector

In [2]:
random.seed(42)

In [3]:
edges_path = 'data/Gowalla_edges.txt'

# Data loading
G = nx.read_edgelist(edges_path, nodetype=int)
checkins_df = pd.read_parquet('data/EDA_output/checkins.parquet')

## Feature Engineering

Initially, a simple random 80/20 split of the edges was considered for creating the training and test sets. However, this approach was identified as methodologically flawed for a temporal dataset. A random split could create a scenario where the model is trained on check-in features from 2010 to predict a link that was held out for the test set, even if that link represents a friendship from 2009. This constitutes a form of data leakage, where information from the future is used to predict the past, leading to an overly optimistic and unrealistic evaluation of the model's performance."

To avoid this temporal leakage and address the limitations of the data, we reframe the prediction task. Given the absence of friendship formation timestamps that makes true temporal prediction impossible, we *shift from predicting link formation to predicting link existence within the final network snapshot*.

> Can we use a user's early activity to effectively distinguish between user pairs who are friends and those who are not in the final, complete social network?

In this new objective we accept **structural leakage** by using the final graph for embeddings, but we explicitly define the task around this by predicting the *final state*, not the *formation*.

### Feature Dataset Train/Test Split

We proceed to make the 80/20 split of the check-ins. We have two alternative:
- Split by volume of check-ins (e.g. train set will have 80% of check-ins)
- Split by time passed (e.g train set will have check-ins that appears in the first 80% of time)

In [4]:
# Split the check-ins dataframe by time
min_ts = checkins_df['check-in_datetime'].min()
max_ts = checkins_df['check-in_datetime'].max()

time_diff = max_ts - min_ts

time_cutoff = min_ts + pd.to_timedelta(0.8 * time_diff)

t_train_df = checkins_df[checkins_df['check-in_datetime'] < time_cutoff]
t_test_df = checkins_df[checkins_df['check-in_datetime'] >= time_cutoff]

print("Check-ins cutoff at: ", time_cutoff)
print(f"Train set size: {len(t_train_df)} - Corresponding percentage of volume: {len(t_train_df)/len(checkins_df)*100:.1f}")
print(f"Test set size:  {len(t_test_df)} - Corresponding percentage of volume: {len(t_test_df)/len(checkins_df)*100:.1f}")

Check-ins cutoff at:  2010-06-20 00:33:12.400000+00:00
Train set size: 2939173 - Corresponding percentage of volume: 45.6
Test set size:  3503690 - Corresponding percentage of volume: 54.4


In [5]:
# Split the check-ins dataframe by volume
volume_cutoff = math.ceil(len(checkins_df) * 0.80)

checkins_df = checkins_df.sort_values(by=['check-in_datetime'])

v_train_df = checkins_df.iloc[:volume_cutoff]
v_test_df = checkins_df.iloc[volume_cutoff:]

volume_cutoff_ts = checkins_df.iloc[volume_cutoff]['check-in_datetime']

print("Check-ins cutoff at: ", volume_cutoff_ts)
print(f"Train set size: {len(v_train_df)} - Corresponding percentage of time: {((volume_cutoff_ts - min_ts) / time_diff) * 100:.1f}")
print(f"Test set size:  {len(v_test_df)} - Corresponding percentage of time: {((max_ts - volume_cutoff_ts) / time_diff) * 100:.1f}")

Check-ins cutoff at:  2010-09-14 14:24:41+00:00
Train set size: 5154291 - Corresponding percentage of time: 93.8
Test set size:  1288572 - Corresponding percentage of time: 6.2


Either way we obtain an overall imbalanced split, and either options have pros and cons:
- Option "Time": This is the most honest representation of a real-world temporal prediction task, but the features built from this sparser data might be weak.
- Option "Volume": The features built will be much stronger and the model will surely perform better, but the separation between past and future is tiny.

As the goal of the notebook is learning, we proceed with the time-based split.

### Negative Sampling

To create a balanced dataset we take an equal number of positive and negative samples. To make the training set challenging, we use a mixed strategy:
- 70% of negative samples are "hard" negatives generated with random walks with a path distance of 2.
- 20% are "medium" negatives generated with random walks with a path distance of 4, representing users who are further apart but still connected.
- 10% are "easy" random negatives, to ensure the model learns the global structure of the graph.

In [6]:
# Negative Sampling Generation Functions
def _random_walk_hard_negative(graph, count, walk_distance=2, max_attempts=100):
    """Generates a batch of hard negative samples."""
    if walk_distance < 1:
        raise ValueError("walk_distance must be >= 1")

    edges = list(graph.edges())
    if not edges:
        return []

    generated_samples = []
    for _ in range(count):
        for _ in range(max_attempts):
            u, current_node = random.choice(edges)

            # Make hops
            for _ in range(walk_distance - 1):
                neighbors = list(graph.neighbors(current_node))
                if not neighbors:
                    current_node = None
                    break
                current_node = random.choice(neighbors)

            if current_node is None:
                continue

            v = current_node

            # Validate edge and add to the batch
            if v != u and not graph.has_edge(u, v):
                generated_samples.append((u, v))
                break

    return generated_samples

def _random_negative_sampling(graph, count, max_attempts=100):
    """Generates a batch of random negative samples."""
    nodes = list(graph.nodes())
    if len(nodes) < 2:
        return []

    generated_samples = []
    for _ in range(count):
        for _ in range(max_attempts):
            u, v = random.sample(nodes, 2)
            if not graph.has_edge(u, v):
                generated_samples.append((u, v))
                break

    return generated_samples

In [7]:
# Main Negative Sampling Function to combine different methods
def generate_negative_samples_set(
    graph,
    sampling_functions,
    sampling_weights,
    ratio=1.0,
    seed=None
):
    if seed is not None:
        random.seed(seed)
        np.random.seed(seed)

    if len(sampling_functions) != len(sampling_weights):
        raise ValueError("sampling_functions and sampling_weights must be of the same length")

    total_neg_samples = int(len(graph.edges) * ratio)

    weights = np.array(sampling_weights, dtype=float)
    weights /= weights.sum()
    samples_for_each_function = np.random.multinomial(total_neg_samples, weights)

    negative_samples = set()

    for func, count in zip(sampling_functions, samples_for_each_function):
        if count == 0:
            continue

        new_samples = func(graph, count)
        negative_samples.update(new_samples)

    return negative_samples

In [8]:
# Generate Negative Samples
positive_samples = list(G.edges)

sampling_functions = [
    lambda g, c: _random_negative_sampling(g, c),
    lambda g, c: _random_walk_hard_negative(g, c, walk_distance=4),
    lambda g, c: _random_walk_hard_negative(g, c, walk_distance=2)
]

sampling_weights = [0.1, 0.2, 0.7]

negative_samples = generate_negative_samples_set(
    graph=G,
    sampling_functions=sampling_functions,
    sampling_weights=sampling_weights,
    ratio=1.1, # Allow to compensate duplicates
    seed=42
)

print(f"Generated {len(negative_samples)} ({len(negative_samples)/len(positive_samples)*100:.2f}%) negative samples.")

Generated 973408 (102.43%) negative samples.


In [9]:
# Truncate the exceeding part
negative_samples = random.sample(list(negative_samples), len(positive_samples))

In [10]:
# Combine and create labels
combined_samples = positive_samples + negative_samples
samples_labels = [1] * len(positive_samples) + [0] * len(negative_samples)

### Node Embeddings

The goal of **Node Embedding** is to learn a *dense vector representation* for each user that captures their neighborhood and role in the network. This approach allows the model to automatically learn features from the graph topology, serving as a powerful alternative to manual feature engineering with standard measures (e.g., Common Neighbors).

We chose the **Node2Vec** algorithm for this task. An initial test with a standard Python implementation of Node2Vec proved to be extremely slow. The first phase of the algorithm, *"computing transition probabilities"* was single-threaded and was projected to take several hours. A Stack Overflow discussion<sup>[[1][1]]</sup> highlighted this common bottleneck and pointed towards more performant, specialized libraries.

[1]: https://stackoverflow.com/questions/60276191/is-there-any-way-to-make-node2vec-faster

Based on this research, we switched to **GRAPE**<sup>[[2][2]]</sup>, a fast graph processing and embedding library. GRAPE is written in *Rust* and *Python*, developed primarily at the University of Milan, and is designed for speed and scalability on massive graphs. It parallelizes the entire Node2Vec process, including the pre-computation step, which dramatically reduces the training time.  
While the library's documentation was found to be in a raw state, we were able to successfully implement it for our purposes.

[2]: https://github.com/AnacletoLAB/grape

To make our workflow efficient and reproducible, we performed this computationally expensive step in a separate script (`train_embeddings.py`) and saved the resulting embeddings to a file. The model was configured with a neutral baseline set of hyperparameters:
- `embedding_size` (dimensions): 128
- `walk_length` (length of each walk): 80
- `iterations` (number of walks per node): 10
- `return_weight` ($p = 1 / \text{return\_weight}$): 1.0
- `explore_weight` ($q = 1 / \text{explore\_weight}$): 1.0

In the Node2Vec framework `return_weight` and `explore_weight` set to `1.0` corresponds to the parameters `p` and `q`, which removes any search bias and makes the algorithm equivalent to the **DeepWalk** algorithm.

The training process, which ran over *30 epochs*, required *35 minutes* to complete.

In [11]:
# Load Embeddings
embedding_path = "embeddings/gowalla_node2vec_embeddings.parquet"
embedding_df = pd.read_parquet(embedding_path)
embedding_df.index = embedding_df.index.astype(int)

In [12]:
# Show result
embedding_df.head()

0           1           2           3          4           5    \
0   -4.028883   -2.877670    5.043583   -0.282113   9.059772    3.420488   
1 -243.156174  -22.406778    8.421268  150.557510 -96.713753  138.841431   
2  -46.896175  -40.126060  116.891319 -351.345337  86.865761  -99.941383   
3  -24.852211  144.911758  -16.660511  143.682907 -99.156891  -38.815899   
4   -1.900526    0.714963   -1.704514    1.345464  -1.145782   -2.345304   

         6          7           8           9    ...         118         119  \
0  -2.221864  -0.936862    4.680723    0.022812  ...    1.390288   -2.516412   
1 -76.178436 -37.150448   77.478683  102.910446  ...  106.663269   32.469856   
2 -94.355743 -35.766766 -103.297546 -206.596237  ...  174.219254  207.147949   
3 -70.038078 -80.872765  -86.241013    5.924940  ...  265.224274   98.895889   
4  -0.432718   0.968551    1.278602   -2.060853  ...   -1.869221    1.397865   

          120         121         122         123         124         125  \
0    6.060315   -0.654452   -0.875411   -1.368012    0.509950   -1.803684   
1   66.770645   84.003662  107.856476  -17.861546 -122.073204 -141.799301   
2 -160.191910  -72.079605  231.860718  160.584747   51.183266    3.251709   
3  248.775223  395.515076  -30.951685 -219.192673  -90.772003  110.593315   
4    0.839844   -1.860224    2.509959   -1.883211    2.112453    1.338032   

          126         127  
0    4.541414    0.619863  
1    8.232184  -47.792175  
2  -20.809055  -19.630239  
3 -107.129738 -313.334290  
4   -2.202100    0.280150  

[5 rows x 128 columns]

In [13]:
print(f'DataFrame with {embedding_df.shape[0]} rows and {embedding_df.shape[1]} columns')

DataFrame with 196591 rows and 128 columns


The next step is to create a single feature vector for each *pair* of users in our sample set. This vector needs to describe the relationship between the two users. The **Hadamard Product** is a simple yet effective method that should capture the interaction and agreement between two users embedding along each dimension.

In [14]:
# Efficient vectorized Hadamard Product
samples_df = pd.DataFrame(combined_samples, columns=['source', 'destination'])
source_vectors = embedding_df.loc[(samples_df['source'])].values
destination_vectors = embedding_df.loc[samples_df['destination']].values
hadamard_product = source_vectors * destination_vectors

In [15]:
# Create Pair Features dataframe with multi-index
pair_features_df = pd.DataFrame(
    hadamard_product,
    index=pd.MultiIndex.from_frame(samples_df),
    columns=[f'embed_hadamard_{i}' for i in range(hadamard_product.shape[1])]
)

In [16]:
# Show result
pair_features_df.head()

embed_hadamard_0  embed_hadamard_1  embed_hadamard_2  \
source destination                                                         
0      1                  979.647766         64.479317         42.473366   
       2                  188.939209        115.469574        589.551086   
       3                  100.126648       -417.008270        -84.028679   
       4                    7.656997         -2.057429         -8.596861   
       5                 -177.432129       -129.863907        180.872009   

                    embed_hadamard_3  embed_hadamard_4  embed_hadamard_5  \
source destination                                                         
0      1                  -42.474213       -876.204529        474.905457   
       2                   99.119041        786.983948       -341.848328   
       3                  -40.534798       -898.338806       -132.769318   
       4                   -0.379573        -10.380519         -8.022085   
       5                    1.508496      -1050.496216        -81.615990   

                    embed_hadamard_6  embed_hadamard_7  embed_hadamard_8  \
source destination                                                         
0      1                  169.258118         34.804832        362.656219   
       2                  209.645630         33.508514       -483.507172   
       3                  155.615082         75.766602       -403.670258   
       4                    0.961441         -0.907399          5.984779   
       5                 -256.680695         16.716686         27.005644   

                    embed_hadamard_9  ...  embed_hadamard_118  \
source destination                    ...                       
0      1                    2.347621  ...          148.292709   
       2                   -4.712930  ...          242.215027   
       3                    0.135161  ...          368.738251   
       4                   -0.047013  ...           -2.598756   
       5                   -3.903668  ...         -168.063080   

                    embed_hadamard_119  embed_hadamard_120  \
source destination                                           
0      1                    -81.707527          404.651123   
       2                   -521.269531         -970.813354   
       3                   -248.862778         1507.656128   
       4                     -3.517605            5.089716   
       5                   -548.348328         -143.000809   

                    embed_hadamard_121  embed_hadamard_122  \
source destination                                           
0      1                    -54.976391          -94.418739   
       2                     47.172665         -202.973404   
       3                   -258.845764           27.095444   
       4                      1.217428           -2.197246   
       5                    -14.381987         -111.514908   

                    embed_hadamard_123  embed_hadamard_124  \
source destination                                           
0      1                     24.434811          -62.251236   
       2                   -219.681870           26.100908   
       3                    299.858215          -46.289188   
       4                      2.576255            1.077245   
       5                    151.453049           94.496552   

                    embed_hadamard_125  embed_hadamard_126  embed_hadamard_127  
source destination                                                              
0      1                    255.761078           37.385757          -29.624620  
       2                     -5.865055          -94.502533          -12.168067  
       3                   -199.475357         -486.520477         -194.224457  
       4                     -2.413386          -10.000648            0.173655  
       5                   -454.180786         -446.752899           50.445763  

[5 rows x 128 columns]

In [17]:
# Free-up memory
del embedding_df
del samples_df
del source_vectors
del destination_vectors
del hadamard_product

### Domain Features

Here is a palette of candidate features that were considered for the link prediction model:
- **Radius of Gyration**: A measure of a user's typical travel radius, distinguishing "stayers" (small radius) from "travelers" (large radius).
- **Total Check-in Count**: The total number of check-ins for a user, indicating their overall activity level.
- **Unique Locations Count**: The number of distinct locations a user has visited, indicating their tendency to explore.
- **Jaccard Similarity of Visited Locations**: The overlap in the set of unique locations visited by two users, measuring shared lifestyle and interests.
- **Co-check-in Count**: The number of times two users checked into the same location within a short time window (e.g., 1 hour), indicating possible real-world interaction.
- **Haversine Distance between User Centroids**: The geographic distance between the average location (centroid) of all check-ins for each user.
- **Haversine Distance between Inferred Home Locations**: The geographic distance between the inferred "home" location of each user.
- **Explicit Topological Features**: Common Neighbors, Jaccard Coefficient, Adamic-Adar, Resource Allocation, Preferential Attachment

The first selection will be based on covering each domain of interaction with one feature that is thought to be the best:
- **Radius of Gyration**: Analyzed in the EDA, represent the travelers vs. stayers concept and is an indicator of heterophily.
- **Haversine Distance between Inferred Home Location**: the core finding in the EDA, represent the geographic proximity and has already been proved a strong indicator of homophily.
- **Jaccard Similarity of Visited Locations**: we didn't explicitly explore this in the EDA, but could strongly represent the shared habits for a pair of users.

In [18]:
checkins_per_user = checkins_df['user'].value_counts()
print(f'[Original Dataset] - Number of users with at least one check-in: {len(checkins_per_user)} ({len(checkins_per_user)/G.number_of_nodes()*100:.2f}%)')
checkins_per_user = t_train_df['user'].value_counts()
print(f'[Time Train Dataset] - Number of users with at least one check-in: {len(checkins_per_user)} ({len(checkins_per_user)/G.number_of_nodes()*100:.2f}%)')
checkins_per_user = v_train_df['user'].value_counts()
print(f'[Volume Train Dataset] - Number of users with at least one check-in: {len(checkins_per_user)} ({len(checkins_per_user)/G.number_of_nodes()*100:.2f}%)')


[Original Dataset] - Number of users with at least one check-in: 107092 (54.47%)
[Time Train Dataset] - Number of users with at least one check-in: 63526 (32.31%)
[Volume Train Dataset] - Number of users with at least one check-in: 94019 (47.82%)


#### Pair-Wise Radius of Gyration

We need to recalculate the Radius of Gyration using only the check-ins before the cutoff date.

In [19]:
# Calculate Radius of Gyration with only check-ins before cutoff
# Calculate Mean Centroid
users_mean_centroids = t_train_df.groupby('user').agg(
    mean_centroid_latitude=('latitude', 'mean'),
    mean_centroid_longitude=('longitude', 'mean')
)

# Merge Mean Centroid with check-ins dataframe
checkins_with_centroids = pd.merge(t_train_df, users_mean_centroids, on='user')
checkins_with_centroids = checkins_with_centroids[['user', 'latitude', 'longitude', 'mean_centroid_latitude', 'mean_centroid_longitude']]

# Zip to obtain coordinates pairs
checkins_coords = list(zip(checkins_with_centroids['latitude'], checkins_with_centroids['longitude']))
centroid_coords = list(zip(checkins_with_centroids['mean_centroid_latitude'], checkins_with_centroids['mean_centroid_longitude']))

# Efficient vectorized Haversine Distance
distances = haversine_vector(checkins_coords, centroid_coords)
checkins_with_centroids['sq_distance_from_centroid'] = distances**2

# Calculate Radius of Gyration and save in a new dataframe
radius_of_gyration = checkins_with_centroids.groupby('user')['sq_distance_from_centroid'].mean().apply(np.sqrt).rename('radius_of_gyration_km')
radius_of_gyration.index = radius_of_gyration.index.astype(int)

In [20]:
# Show results
radius_of_gyration.head()

user
0     1203.823612
4      421.415229
5       49.801932
9        5.595743
10       8.475587
Name: radius_of_gyration_km, dtype: float64

In [21]:
# Calculate the pair-wise feature: Radius of Gyration Absolute Difference
samples_df = pd.DataFrame(combined_samples, columns=['source', 'destination'])
source_rg_vector = samples_df['source'].map(radius_of_gyration)
destination_rg_vector = samples_df['destination'].map(radius_of_gyration)
abs_diff_rg = (source_rg_vector - destination_rg_vector).abs()

rg_feature_df = pd.DataFrame({
    'rg_source': source_rg_vector,
    'rg_destination': destination_rg_vector,
    'rg_abs_diff': abs_diff_rg
})
rg_feature_df.index = pd.MultiIndex.from_frame(samples_df)

In [22]:
# Check for NaN values
print(rg_feature_df.isna().sum(), "\n")

rg_na = rg_feature_df.isna().sum().loc['rg_abs_diff']
print(f"Radius of Gyration absolute difference missing for {rg_na} ({rg_na/len(combined_samples)*100:.4}%) pairs.")

rg_source          991583
rg_destination    1110422
rg_abs_diff       1470012
dtype: int64 

Radius of Gyration absolute difference missing for 1470012 (77.34%) pairs.


The **Radius of Gyration (RoG)** was calculated for all users based on their activity before the time-based cutoff date. As expected, and as a direct consequence of the network's growth, only 32.3% of the total users had check-in data within this period.

We now need to deal with the presence of numerous missing values due to having many users not active on the platform in this time range. To treat this information as a predictive signal rather than a data problem, we engineered a dedicated categorical feature, `activity_status`, to explicitly describe the activity state of each pair (`Both Active`, `Source Active Only`, etc.). This allows the model to learn distinct patterns for each scenario. With a feature that gives context the the RoG values, all the NaN can now be filled with zeros. 

In [23]:
# Fill the NaN values
rg_feature_df['rg_source'] = rg_feature_df['rg_source'].fillna(0)
rg_feature_df['rg_destination'] = rg_feature_df['rg_destination'].fillna(0)

rg_feature_df['rg_abs_diff'] = (rg_feature_df['rg_source'] - rg_feature_df['rg_destination']).abs()

In [24]:
# Add categorical feature for context
source_is_active = samples_df['source'].isin(radius_of_gyration.index)
destination_is_active = samples_df['destination'].isin(radius_of_gyration.index)

conditions = [
    source_is_active & destination_is_active,
    source_is_active & ~destination_is_active,
    ~source_is_active & destination_is_active,
    ~source_is_active & ~destination_is_active
]
choices = ['both_have', 'source_only', 'dest_only', 'neither_have']

activity_status = np.select(conditions, choices, default='Error')

rg_feature_df['activity_status'] = activity_status
rg_feature_df.head()

rg_source  rg_destination  rg_abs_diff activity_status
source destination                                                          
0      1            1203.823612        0.000000  1203.823612     source_only
       2            1203.823612        0.000000  1203.823612     source_only
       3            1203.823612        0.000000  1203.823612     source_only
       4            1203.823612      421.415229   782.408383       both_have
       5            1203.823612       49.801932  1154.021681       both_have

In [25]:
# One-hot encoding categorical feature
rg_feature_df = pd.get_dummies(rg_feature_df, columns=['activity_status'], prefix='rg_status')
rg_feature_df.head()

rg_source  rg_destination  rg_abs_diff  \
source destination                                             
0      1            1203.823612        0.000000  1203.823612   
       2            1203.823612        0.000000  1203.823612   
       3            1203.823612        0.000000  1203.823612   
       4            1203.823612      421.415229   782.408383   
       5            1203.823612       49.801932  1154.021681   

                    rg_status_both_have  rg_status_dest_only  \
source destination                                             
0      1                          False                False   
       2                          False                False   
       3                          False                False   
       4                           True                False   
       5                           True                False   

                    rg_status_neither_have  rg_status_source_only  
source destination                                                 
0      1                             False                   True  
       2                             False                   True  
       3                             False                   True  
       4                             False                  False  
       5                             False                  False

In [26]:
# Add results to the complete pair feature dataframe
pair_features_df = pair_features_df.join(rg_feature_df.drop(columns=['rg_source', 'rg_destination']))

In [27]:
# Free-Up Memory
del users_mean_centroids
del checkins_with_centroids
del checkins_coords
del centroid_coords
del distances
del radius_of_gyration
del samples_df
del source_rg_vector
del destination_rg_vector
del abs_diff_rg
del rg_feature_df
del source_is_active
del destination_is_active
del activity_status

#### Haversine Distance between Inferred Home Locations

Again, all the calculations needs to be redone using only the check-ins before the cutoff date.

In [ ]:
# Reuse code from EDA
inferred_home_df = t_train_df.groupby('user').agg(
    median_centroid_latitude=('latitude', 'median'),
    median_centroid_longitude=('longitude', 'median')
)

# Heuristic to infer Home Locations
home_hours = (t_train_df['check-in_datetime'].dt.hour >= 21) | (t_train_df['check-in_datetime'].dt.hour < 7)

# The grid will have 25x25 Kilometers cells
lat_step, lon_step = 0.25, 0.25
t_train_df['lat_bin'] = (t_train_df['latitude'] / lat_step).astype(int)
t_train_df['lon_bin'] = (t_train_df['longitude'] / lon_step).astype(int)

# Find the most visited grid cell for each user during home hours
home_cells = t_train_df[home_hours]\
    .groupby(['user', 'lat_bin', 'lon_bin'])\
    .size()\
    .reset_index(name='count')\
    .sort_values('count', ascending=False)\
    .drop_duplicates(subset='user')\
    .set_index('user')\
    [['lat_bin', 'lon_bin']]\
    .rename(columns={'lat_bin': 'home_lat_bin', 'lon_bin': 'home_lon_bin'})

# Chain operations to calculate the centroid of check-ins within each user's home cell
home_cell_centroids = (
    t_train_df.join(home_cells, on='user')
    .dropna(subset=['home_lat_bin'])
    .query('lat_bin == home_lat_bin and lon_bin == home_lon_bin')
    .groupby('user')
    .agg(
        home_cell_centroid_lat=('latitude', 'median'),
        home_cell_centroid_lon=('longitude', 'median')
    )
)

inferred_home_df = inferred_home_df.join(home_cell_centroids)
inferred_home_df.index = inferred_home_df.index.astype(int)
inferred_home_df = inferred_home_df[['home_cell_centroid_lat', 'home_cell_centroid_lon']]

In [29]:
inferred_home_df.isna().sum()

home_cell_centroid_lat    10739
home_cell_centroid_lon    10739
dtype: int64

In [30]:
samples_df = pd.DataFrame(combined_samples, columns=["source", "destination"])

samples_df = samples_df.join(inferred_home_df, on="source")
samples_df = samples_df.rename(columns={
    "home_cell_centroid_lat": "source_home_lat",
    "home_cell_centroid_lon": "source_home_lon",
})

samples_df = samples_df.join(inferred_home_df, on="destination")
samples_df = samples_df.rename(columns={
    "home_cell_centroid_lat": "dest_home_lat",
    "home_cell_centroid_lon": "dest_home_lon",
})

In [31]:
samples_df.index = pd.MultiIndex.from_frame(samples_df[["source", "destination"]])
samples_df = samples_df.drop(columns=["source", "destination"])

In [32]:
source_home_coords = list(zip(samples_df['source_home_lat'], samples_df['source_home_lon']))
dest_home_coords = list(zip(samples_df['dest_home_lat'], samples_df['dest_home_lon']))

distances = haversine_vector(source_home_coords, dest_home_coords)
samples_df['home_distance_km'] = distances

In [33]:
# Check for NaN values
print(samples_df.isna().sum(), "\n")

home_dist_na = samples_df.isna().sum().loc['home_distance_km']
print(f"Home Distance missing for {home_dist_na} ({home_dist_na/len(combined_samples)*100:.4}%) pairs.")

source_home_lat     1060573
source_home_lon     1060573
dest_home_lat       1201301
dest_home_lon       1201301
home_distance_km    1539456
dtype: int64 

Home Distance missing for 1539456 (81.0%) pairs.


In [34]:
source_has_home = samples_df['source_home_lat'].notna()
dest_has_home = samples_df['dest_home_lat'].notna()

conditions = [
    source_has_home & dest_has_home,
    source_has_home & ~dest_has_home,
    ~source_has_home & dest_has_home,
    ~source_has_home & ~dest_has_home
]

choices = ['both_have', 'source_only', 'dest_only', 'neither_have']
samples_df['home_status'] = np.select(conditions, choices)

The `home_distance_km` column now contains the calculated distances for pairs where both users had an inferred home, and NaN for all other pairs. A simple imputation with zero would be misleading, as it would imply that users with no inferred home live at the same location. Therefore, a more neutral and statistically robust approach is chosen: the missing distance values are imputed using the **median** of all the calculated distances.

In [35]:
# Impute the missing home distance values with the median of the distances
median_distance = samples_df['home_distance_km'].median()
samples_df['home_distance_km'] = samples_df['home_distance_km'].fillna(median_distance)

print("Median Home Distance: ", median_distance)

Median Home Distance:  858.9177190455824


In [36]:
# One-hot encoding categorical feature
samples_df = pd.get_dummies(samples_df, columns=['home_status'], prefix='home_status')

In [38]:
# Add results to the complete pair feature dataframe
pair_features_df = pair_features_df.join(samples_df.drop(columns=['source_home_lat', 'source_home_lon', 'dest_home_lat', 'dest_home_lon']))

In [39]:
# Free-Up Memory
del inferred_home_df
del home_cells
del home_hours
del home_cell_centroids
del samples_df
del source_home_coords
del dest_home_coords

#### Jaccard Similarity of Visited Locations